sync/fetch_queue.go

```diff
@@ -0,0 +1,315 @@
+package sync
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"sync"
+)
+
+type FetchPoetProofFunc func(poetProofRef []byte) error
+type GetValidAddressableTxFunc func(tx *types.SerializableSignedTransaction) (*types.AddressableSignedTransaction, error)
+type SValidateAtxFunc func(atx *types.ActivationTx) error
+type CheckLocalFunc func(ids []types.Hash32) (map[types.Hash32]Item, map[types.Hash32]Item, []types.Hash32)
+
+type Item interface {
+	Hash32() types.Hash32
+	ShortString() string
+}
+
+type fetchJob struct {
+	items []Item
+	ids   []types.Hash32
+}
+
+//todo make the queue generic
+type fetchQueue struct {
+	log.Log
+	BatchRequestFactory
+	*sync.Mutex
+	workerInfra WorkerInfra
+	pending     map[types.Hash32][]chan bool
+	handle      func(fj fetchJob)
+	checkLocal  CheckLocalFunc
+	queue       chan []types.Hash32 //types.TransactionId //todo make buffered
+}
+
+func (vq *fetchQueue) Close() {
+	vq.Info("done")
+	close(vq.queue)
+}
+
+//todo batches
+func (fq *fetchQueue) work() error {
+	output := fetchWithFactory(NewFetchWorker(fq.workerInfra, 1, fq.BatchRequestFactory, fq.queue))
+	for out := range output {
+		if out == nil {
+			fq.Info("close queue")
+			return nil
+		}
+
+		bjb, ok := out.(fetchJob)
+		if !ok {
+			fq.Error(fmt.Sprintf("Type conversion err %v", out))
+			continue
+		}
+
+		if len(bjb.ids) == 0 {
+			return fmt.Errorf("channel closed")
+		}
+
+		if len(bjb.items) == 0 {
+			fq.Warning("could not fetch any items ") //todo print ids
+			continue
+		}
+
+		fq.handle(bjb)
+		fq.Debug("next batch")
+	}
+	return nil
+}
+
+func (fq *fetchQueue) addToPendingGetCh(ids []types.Hash32) chan bool {
+	return getDoneChan(fq.addToPending(ids))
+}
+
+func (fq *fetchQueue) addToPending(ids []types.Hash32) []chan bool {
+	fq.Lock()
+	deps := make([]chan bool, 0, len(ids))
+	var idsToAdd []types.Hash32
+	for _, id := range ids {
+		ch := make(chan bool, 1)
+		deps = append(deps, ch)
+		if _, ok := fq.pending[id]; !ok {
+			idsToAdd = append(idsToAdd, id)
+		}
+		fq.pending[id] = append(fq.pending[id], ch)
+	}
+	fq.Unlock()
+	fq.queue <- idsToAdd
+	return deps
+}
+
+func (fq *fetchQueue) invalidate(id types.Hash32, valid bool) {
+	fq.Debug("done with %v !!!!!!!!!!!!!!!! %v", id.ShortString(), valid)
+	fq.Lock()
+	deps := fq.pending[id]
+	delete(fq.pending, id)
+	fq.Unlock()
+
+	for _, dep := range deps {
+		dep <- valid
+	}
+}
+
+//returns items out of itemIds that are not in the local database
+func (tq *fetchQueue) Handle(itemIds []types.Hash32) ([]Item, error) {
+	if len(itemIds) == 0 {
+		tq.Debug("handle empty item ids slice")
+		return nil, nil
```

Comment:
why not ? because it's an interface ? its ok
but i will add an error  to go along with it

Reply:
I saw that in some cases the caller if `Handle` calles `len` and `range` on the resulted map. I thought that it is a problem in case the map is `nil` but I was wrong. I do think that an empty result is nicer than a nil result, but however you prefer